src/caip-types.ts

```diff
@@ -0,0 +1,191 @@
+import type { Infer } from 'superstruct';
+import {
+  array,
+  is,
+  object,
+  optional,
+  pattern,
+  size,
+  string,
+} from 'superstruct';
+
+export const CAIP_CHAIN_ID_REGEX =
+  /^(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-a-zA-Z0-9]{1,32})$/u;
+
+export const CAIP_NAMESPACE_ID_REGEX = /^[-a-z0-9]{3,8}$/u;
+
+export const CAIP_ACCOUNT_ID_REGEX =
+  /^(?<chainId>(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-a-zA-Z0-9]{1,32})):(?<accountAddress>[-.%a-zA-Z0-9]{1,128})$/u;
+
+export const CAIP_ACCOUNT_ADDRESS_REGEX =
+  /^(?<accountAddress>[-.%a-zA-Z0-9]{1,128})$/u;
+
+/**
+ * Parse a chain ID string to an object containing the namespace and reference.
+ * This validates the chain ID before parsing it.
+ *
+ * @param caipChainId - The chain ID to validate and parse.
+ * @returns The parsed chain ID.
+ */
+export function parseCaipChainId(caipChainId: CaipChainId): {
+  namespace: CaipNamespaceId;
+  reference: string;
+} {
+  const match = CAIP_CHAIN_ID_REGEX.exec(caipChainId);
+  if (!match?.groups) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  return {
+    namespace: match.groups.namespace as CaipNamespaceId,
+    reference: match.groups.reference as string,
+  };
+}
+
+/**
+ * Parse an account ID to an object containing the chain, chain ID and address.
+ * This validates the account ID before parsing it.
+ *
+ * @param accountId - The account ID to validate and parse.
+ * @returns The parsed account ID.
+ */
+export function parseCaipAccountId(accountId: CaipAccountId): {
+  chain: { namespace: CaipNamespaceId; reference: string };
+  chainId: CaipChainId;
+  address: string;
+} {
+  const match = CAIP_ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress as string,
+    chainId: match.groups.chainId as CaipChainId,
+    chain: {
+      namespace: match.groups.namespace as CaipNamespaceId,
+      reference: match.groups.reference as string,
+    },
+  };
+}
+
+/**
+ * A helper struct for a string with a minimum length of 1 and a maximum length
+ * of 40.
+ */
+export const LimitedString = size(string(), 1, 40);
```

Comment:
any idea why we are doing this?

Reply:
It's used in `ChainStruct` to limit the chain name, not sure what the motivation was behind 40, specifically.